src/hooks/useStableQuery.ts

```diff
@@ -0,0 +1,56 @@
+import { useRef } from 'react'
+import { useQuery, usePaginatedQuery } from '../convex/_generated/react'
+
+/**
+ * Drop-in replacement for useQuery intended to be used with a parametrized query.
+ * Unlike useQuery, useStableQuery does not return undefined while loading new
+ * data when the query arguments change, but instead will continue to return
+ * the previously loaded data until the new data has finished loading.
+ *
+ * See stack.convex.dev post "Help, my data is overreacting!" for details.
+ * 
+ * @param name - string naming the query function
+ * @param ...args - any number of arguments to be passed to the query function
+ * @returns UseQueryResult
+ */
+export const useStableQuery = ((name, ...args) => {
+  const result = useQuery(name, ...args)
+  const stored = useRef(result) // ref objects are stable between rerenders
+
+  // result is only undefined while data is loading
+  // if a freshly loaded result is available, use the ref to store it
+  if (result !== undefined) {
+    stored.current = result
+  }
+
+  // undefined on first load, stale data while loading, fresh data after loading
+  return stored.current 
+}) as typeof useQuery
+
+
+/**
+ * Drop-in replacement for usePaginatedQuery for use with a parametrized query.
+ * Unlike usePaginatedQuery, when query arguments change useStablePaginatedQuery 
+ * does not return empty results and 'LoadingMore' status. Instead, it continues
+ * to return the previously loaded results until the new results have finished
+ * loading.
+ *
+ * See stack.convex.dev post "Help, my data is overreacting!" for details.
```

Comment:
You can point at https://stack.convex.dev/help-my-app-is-overreacting btw - it'll be available once I publish it, but not on the homepage until after the specified date